geomstats/stiefel.py

```diff
@@ -0,0 +1,351 @@
+"""
+Stiefel manifold St(n,p),
+a set of all orthonormal p-frames in n-dimensional space,
+where p <= n
+"""
+
+import geomstats.backend as gs
+
+from geomstats.embedded_manifold import EmbeddedManifold
+from geomstats.riemannian_metric import RiemannianMetric
+
+TOLERANCE = 1e-6
+EPSILON = 1e-6
+
+
+def matrix_f(sq_mat, f):
+
+    sq_mat = gs.to_ndarray(sq_mat, to_ndim=3)
+
+    [eigenvalues, vectors] = gs.linalg.eig(sq_mat)
+
+    exp_eigenvalues = f(eigenvalues)
+
+    aux = gs.einsum('ijk,ik->ijk', vectors, exp_eigenvalues)
+    exp_mat = gs.einsum('ijk,ikl->ijl', aux, gs.linalg.inv(vectors))
+
+    return exp_mat.real
+
+
+class Stiefel(EmbeddedManifold):
+
+    def __init__(self, n, p):
+        assert(n >= p)
+
+        self.n = n
+        self.p = p
+
+        self.dimension = int(p * n - (p * (p + 1) / 2))
+
+    def belongs(self, point, tolerance=TOLERANCE):
+        """
+        Evaluate if a point belongs to St(n,p),
+        i.e. if it is a p-frame in n-dimensional space,
+        and it is orthonormal.
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        n_points, n, p = point.shape
+
+        if (n, p) != (self.n, self.p):
+            return gs.zeros((n_points,)).astype(bool)
+
+        diff = gs.matmul(gs.transpose(point, axes=(0, 2, 1)), point) - gs.eye(p)
+        point_norm = gs.norm(diff, axis=(1, 2))
+
+        return gs.less_equal(point_norm, tolerance)
+
+    def project(self):
+        pass
+
+    def random_uniform(self, n_samples=1):
+        """
+        Sample on St(n,p) with the uniform distribution.
+
+        If Z(p,n) ~ N(0,1), then St(n,p) ~ U, according to Haar measure:
+        St(n,p) := Z(Z^TZ)^{-1/2}
+        """
+        Z = gs.random.normal(shape=(n_samples, self.n, self.p))
+        return gs.matmul(Z, gs.linalg.inv(
```

Comment:
could you decompose these operations on several lines to improve readability? thank you!